File: src/graphs/number-of-islands-200.py
class Solution(object):
    def numIslands(self, grid):
        if len(grid) == 0 or len(grid[0]) == 0:
            return 0
        directions = [
            [0, 1],
            [1, 0],
            [0, -1],
            [-1, 0]
        ]
        stack = []
        result = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == "1":
                    stack.append([i, j])
                    grid[i][j] = 0
                    self.bfsGraph(directions, grid, stack)
                    result += 1
        return result

    def bfsGraph(self, directions, grid, stack):
        while len(stack) > 0:
            node = stack.pop()
            for direction in directions:
                if self.inBound(node, direction, grid):
                    stack.append([node[0] + direction[0], node[1] + direction[1]])
                    grid[node[0] + direction[0]][node[1] + direction[1]] = "0"

    def inBound(self, node, direction, grid):
        row = node[0] + direction[0]
        column = node[1] + direction[1]
        return len(grid) > row >= 0 and len(grid[0]) > column >= 0 and grid[row][column] == "1"

File: src/depth-first-searches/number-of-islands-200.java
class Solution {
    public int numIslands(char[][] grid) {
        if(grid.length == 0 || grid[0].length == 0)
            return 0;
        int result = 0;
        
        //VISTED ISLAND = 0
        
        for(int i = 0; i < grid.length; i++){
            for(int j = 0; j < grid[0].length; j++){
                if(grid[i][j] == '1'){
                    result += 1;
                    dfs(grid, i, j);
                }
            }
        }
        
        return result;
    }
    
    public void dfs(char[][] grid, int i, int j){
        if(i < 0 || i >= grid.length || j < 0 || j >= grid[0].length || grid[i][j] == '0')
            return;
        
        grid[i][j] = '0';
        dfs(grid, i + 1, j);
        dfs(grid, i - 1, j);
        dfs(grid, i, j + 1);
        dfs(grid, i, j - 1);
    }
}